### PR TITLE
Feat/kbqa release

### PR DESCRIPTION
**Pull request by dmitrijeuseew**
Knowledge base based Question Answering is added.
This system is based on the existing `NER` and `Classification` Components along with two new components:
`Entity Linker` - for linking extracted entity with instance in the knowledge base
`KB Answer Parser` - for aggregation of results of the models mentioned above and generation of the answer in natural language.
The system is trained to answer factoid questions in Russian.

### CODE REVIEW

**Comment by mu-arkhipov** [comment]
Remove all KBQA configs except the final one

**Comment by mu-arkhipov** [comment]
Put this config to the tests

File: deeppavlov/models/kbqa/entity_linking.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from pathlib import Path
+
+import pickle
+from fuzzywuzzy import fuzz
+import pymorphy2
+import itertools
```
Comment (yoptar) [inline deeppavlov/models/kbqa/entity_linking.py]: imports order

File: deeppavlov/models/kbqa/kb_answer_parser_wikidata.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from logging import getLogger
+from typing import List, Tuple
+
+import numpy as np
+import pickle
+from deeppavlov.core.models.serializable import Serializable
```
Comment (yoptar) [inline deeppavlov/models/kbqa/kb_answer_parser_wikidata.py]: imports order

File: deeppavlov/models/kbqa/entity_linking.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+from typing import List, Dict, Tuple
+
+from deeppavlov.core.common.registry import register
+from deeppavlov.core.models.serializable import Serializable
+
+log = getLogger(__name__)
+
+
+@register('entity_linker')
+class EntityLinker(Serializable):
```
Comment (yoptar) [inline deeppavlov/models/kbqa/entity_linking.py]:
```suggestion
class EntityLinker(Component, Serializable):
```

File: deeppavlov/models/kbqa/entity_linking.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        self.morph = pymorphy2.MorphAnalyzer()
+        self.lemmatize = lemmatize
+        self.debug = debug
+        self.rule_filter_entities = rule_filter_entities
+        self.substring_entity_search = substring_entity_search
+
+        self._wiki_filename = wiki_filename
+        self._entities_filename = entities_filename
+
+        self.name_to_q = None
```
Comment (yoptar) [inline deeppavlov/models/kbqa/entity_linking.py]:
```python3
self.name_to_q: Optional[Dict[str, List[Tuple[str]]]] = None
```

File: deeppavlov/models/kbqa/entity_linking.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+    def _log_entities(self, srtd_cand_ent):
+        entities_to_print = []
+        for name, q, n_rel in srtd_cand_ent:
+            entities_to_print.append(f'{name}, http://wikidata.org/wiki/{q[1]}, {n_rel[2]}')
+        log.debug('\n'.join(entities_to_print))
+
+    def find_candidate_entities(self, entity: str) -> List[str]:
+        candidate_entities = []
+        candidate_entities += self.name_to_q.get(entity, [])
```
Comment (yoptar) [inline deeppavlov/models/kbqa/entity_linking.py]: If you want to copy a list
```python3
candidate_entities = list(self.name_to_q.get(entity, []))
```

File: deeppavlov/models/kbqa/entity_linking.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+                entity_triplets.append([])
+
+        return entity_triplets
+
+    @staticmethod
+    def filter_triplets(entity_triplets: List[List[List[str]]], question_tokens: List[str],
+        srtd_cand_ent: List[Tuple[str]]) -> List[List[List[str]]]:
+
+        question_begin = question_tokens[0].lower()
+        what_template = 'что'
```
Comment (yoptar) [inline deeppavlov/models/kbqa/entity_linking.py]: Can this class be only used for Russian?

File: deeppavlov/models/kbqa/kb_answer_parser_wikidata.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+        self._templates_filename = templates_filename
+        self._q_to_name = None
+        self._relations_mapping = None
+        self.templates = None
+        self.return_confidences = return_confidences
+        self.linker = linker
+        self.load()
+
+    def load(self) -> None:
+        load_path = Path(self.load_path).expanduser()
```
Comment (yoptar) [inline deeppavlov/models/kbqa/kb_answer_parser_wikidata.py]: This should already be done in `Serializable.__init__()`

File: docs/components/kbqa.rst [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+Models can be used from Python using the following code:
+
+.. code:: python
+
+    from deeppavlov import configs, build_model
+
+    kbqa_model = build_model(configs.kbqa.kbqa_mix_lowercase, download=True)
+    kbqa_model(['Когда родился Пушкин?'])
+    >>> ["26 May 1799"]
```
Comment (yoptar) [inline docs/components/kbqa.rst]: Can we output dates in Russian for Russian model?

File: requirements.txt [hunk 9]
```diff
@@ -11,8 +11,9 @@ scikit-learn==0.19.1
 fuzzywuzzy==0.16.0
 pymorphy2==0.8
 pymorphy2-dicts-ru
-requests==2.19.1
-pytelegrambotapi==3.5.2
+requests[socks]==2.10.0
```
Comment (yoptar) [inline requirements.txt]: This requests version breaks tests for now

File: deeppavlov/dataset_readers/kbqa_reader.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        data_path = Path(data_path)
+        files = list(data_path.glob('*.txt'))
+        if 'test_set_with_answers.txt' not in {file_path.name for file_path in files}:
+            url = 'http://files.deeppavlov.ai/kbqa/test_set_with_answers.txt'
+            data_path.mkdir(exist_ok=True, parents=True)
+            download_decompress(url, data_path)
+            files = list(data_path.glob('*.txt'))
+        dataset = {}
+
+        dataset["test"] = self.parse_ner_file(files[0])
```
Comment (yoptar) [inline deeppavlov/dataset_readers/kbqa_reader.py]: Are you sure you want to use `files[0]` and not just `data_path / 'test_set_with_answers.txt'`?

File: deeppavlov/metrics/accuracy.py [hunk 11]
```diff
@@ -106,3 +106,13 @@ def round_accuracy(y_true, y_predicted):
     examples_len = len(y_true)
     correct = sum([y1 == y2 for y1, y2 in zip(y_true, predictions)])
     return correct / examples_len if examples_len else 0
+
+
+@register_metric('kbqa_accuracy')
+def kbqa_accuracy(y_true, y_predicted):
+    total_correct = 0
+    for answer_true, answer_predicted in zip(y_true, y_predicted):
+        if answer_predicted in list(answer_true):
```
Comment (yoptar) [inline deeppavlov/metrics/accuracy.py]:
```suggestion
        if answer_predicted in answer_true:
```
Are you sure you need casting here?

File: deeppavlov/configs/kbqa/kbqa_rus.json [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+      {
+        "in": ["x_tokens", "y_predicted", "y_pred_probas"],
+        "out": ["answers"],
+        "load_path": "{DOWNLOADS_PATH}/wikidata_rus/wiki_rus_q_to_name.pickle",
+        "class_name": "kb_answer_parser_wikidata",
+        "classes_vocab_keys": "#classes_vocab.keys()",
+        "top_k_classes": 5,
+        "relations_maping_filename": "wiki_properties_id_to_name.pickle",
+        "templates_filename": "templates.pickle",
+        "debug": true,
```
Comment (yoptar) [inline deeppavlov/configs/kbqa/kbqa_rus.json]:
```suggestion
        "debug": false,
```

File: deeppavlov/models/kbqa/kb_answer_parser_wikidata.py [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+            is_kbqa = self.is_kbqa_question(tokens)
+            if is_kbqa:
+                if self._templates_filename is not None:
+                    entity_from_template, relation_from_template = self.entities_and_rels_from_templates(tokens)
+                else:
+                    entity_from_template = None
+                if entity_from_template:
+                    if self._debug:
+                        relation_title = self._relations_mapping[relation_from_template]
+                        log.info("entity {}, relation {}".format(entity_from_template, relation_title))
```
Comment (yoptar) [inline deeppavlov/models/kbqa/kb_answer_parser_wikidata.py]:
```suggestion
                        log.debug("entity {}, relation {}".format(entity_from_template, relation_title))
```

File: deeppavlov/models/kbqa/kb_answer_parser_wikidata.py [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+                                                          entity_linking_confidences,
+                                                          [relation_from_template],
+                                                          [relation_prob])
+                else:
+                    entity_from_ner = self.extract_entities(tokens, tags)
+                    entity_triplets, entity_linking_confidences = self.linker(entity_from_ner, tokens)
+                    top_k_relations, top_k_probs = self._parse_relations_probs(relations_probs)
+                    top_k_relation_names = [self._relations_mapping[rel] for rel in top_k_relations]
+                    if self._debug:
+                        log.info("top k relations {}" .format(str(top_k_relation_names)))
```
Comment (yoptar) [inline deeppavlov/models/kbqa/kb_answer_parser_wikidata.py]:
```suggestion
                        log.debug("top k relations {}" .format(str(top_k_relation_names)))
```

File: deeppavlov/configs/kbqa/kbqa_rus.json [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+        "in": ["x_tokens", "y_predicted", "y_pred_probas"],
+        "out": ["answers"],
+        "load_path": "{DOWNLOADS_PATH}/wikidata_rus/wiki_rus_q_to_name.pickle",
+        "class_name": "kb_answer_parser_wikidata",
+        "classes_vocab_keys": "#classes_vocab.keys()",
+        "top_k_classes": 5,
+        "relations_maping_filename": "wiki_properties_id_to_name.pickle",
+        "templates_filename": "templates.pickle",
+        "debug": false,
+        "return_confidences": false,
```
Comment (yoptar) [inline deeppavlov/configs/kbqa/kbqa_rus.json]: Why do you need this option? It does not affect computation times nor memory usage and you can just ignore confidences afterwards

File: deeppavlov/models/kbqa/kb_answer_parser_wikidata.py [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+                                       "что делать", "зачем", "что может"]
+        kbqa_question_templates = ["как зовут", "как называется", "как звали", "как ты думаешь", "как твое мнение", \
+                                   "как ты считаешь"]
+        question_init = ' '.join(question_tokens)
+        question = ''.join([ch for ch in question_init if ch not in punctuation]).lower()
+        is_kbqa = True
+        for template in not_kbqa_question_templates:
+            if question.find(template) > -1:
+                is_kbqa = False
+        for template in kbqa_question_templates:
```
Comment (yoptar) [inline deeppavlov/models/kbqa/kb_answer_parser_wikidata.py]:
```python3
is_kbqa = (all(template not in question for template in not_kbqa_question_templates) or
           all(template in question for template in kbqa_question_templates))
```

**Pull request merged**